### PR TITLE
[Heap Trace Standalone] do not allocate memory until init is called (IDFGH-10247)

### PR DESCRIPTION
**Pull request by chipweinberger**
ESP-IDF convention is to not allocate memory until init is called by the user.

This also removes the requirement for `SPIRAM_ALLOW_BSS_SEG_EXTERNAL_MEMORY`

### CODE REVIEW

File: components/heap/Kconfig [hunk 1]
```diff
@@ -75,34 +75,28 @@ menu "Heap memory debugging"
...
-            Enable this flag to use a hash map to increase performance in handling
-            heap trace records.
+            Heap trace standalone supports storing records either as a list or a list + hash map.
 
-            Keeping this as "n" in your project will save RAM and heap memory but will lower
-            the performance of the heap trace in adding, retrieving and removing trace records.
-            Making this as "y" in your project, you will decrease free RAM and heap memory but,
-            the heap trace performances in adding retrieving and removing trace records will be
-            enhanced.
+            Using only a list takes less memory, but calls to 'free' will slow as the
```
Comment (KaeLL) [inline components/heap/Kconfig:80]: s/slow/get\ slower/

**Review by ESP-Marius — COMMENTED** [review]
Looks like a reasonable change to me, just left two minor comments.

File: components/heap/heap_trace_standalone.c [hunk 2]
```diff
@@ -140,6 +135,19 @@ esp_err_t heap_trace_init_standalone(heap_trace_record_t *record_buffer, size_t
         return ESP_ERR_INVALID_ARG;
     }
 
+#if CONFIG_HEAP_TRACE_HASH_MAP
+    if (hash_map == NULL) {
+        uint32_t map_size = sizeof(heap_trace_record_list_t) * CONFIG_HEAP_TRACE_HASH_MAP_SIZE;
+#if CONFIG_HEAP_TRACE_HASH_MAP_IN_EXT_RAM
+        esp_rom_printf("[Heap Trace] hash map: allocating %" PRIu32 " bytes (PSRAM)\n", map_size);
```
[the remark this answers is not in the document]
Comment (chipweinberger) [reply components/heap/heap_trace_standalone.c:142]: I can't recall to be honest.

This is the heap component, is the log component included?

[the remark this answers is not in the document]
Comment (ESP-Marius) [reply components/heap/heap_trace_standalone.c:142]: Yeah, using it here should be fine. You might just need to add a `TAG` if this is the first time it is used.

**Comment by chipweinberger** [comment]
updated.

**Comment by ESP-Marius** [comment]
Thanks! 

I'll sync it to our internal gitlab then.

**Comment by ESP-Marius** [comment]
sha=fc1e8b036546326c2ece3d1d498c6155eb85171e

**Comment by chipweinberger** [comment]
@ESP-Marius bump

**Pull request merged**